File: IAIso-v5.0/core/iaiso-python/iaiso/middleware/openai.py
"""Middleware for the OpenAI Python SDK (v1.x).

Wraps `openai.OpenAI` and `openai.AsyncOpenAI` so that every chat completion
call is accounted for. Tokens are read from `response.usage`; tool calls are
counted from `choice.message.tool_calls`.

Install:  pip install iaiso[openai]
"""

from __future__ import annotations

from typing import Any

from iaiso.core import BoundedExecution, StepOutcome


class EscalationRaised(RuntimeError):
    """Raised when the execution has reached escalation and
    `raise_on_escalation=True` was set.
    """


class OpenAIBoundedClient:
    """Wraps an OpenAI v1 client for pressure accounting.

    Example:
        >>> from openai import OpenAI
        >>> from iaiso import BoundedExecution
        >>> from iaiso.middleware.openai import OpenAIBoundedClient
        >>>
        >>> raw = OpenAI()
        >>> with BoundedExecution.start() as exec:
        ...     client = OpenAIBoundedClient(raw, exec)
        ...     resp = client.chat.completions.create(
        ...         model="gpt-4o",
        ...         messages=[{"role": "user", "content": "hi"}],
        ...     )
    """

    def __init__(
        self,
        client: Any,
        execution: BoundedExecution,
        *,
        raise_on_escalation: bool = False,
    ) -> None:
        self._client = client
        self._execution = execution
        self._raise_on_escalation = raise_on_escalation
        self.chat = _ChatProxy(client.chat, execution, raise_on_escalation)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._client, name)


class _ChatProxy:
    def __init__(
        self,
        inner: Any,
        execution: BoundedExecution,
        raise_on_escalation: bool,
    ) -> None:
        self._inner = inner
        self._execution = execution
        self._raise_on_escalation = raise_on_escalation
        self.completions = _CompletionsProxy(
            inner.completions, execution, raise_on_escalation
        )

    def __getattr__(self, name: str) -> Any:
        return getattr(self._inner, name)


class _CompletionsProxy:
    def __init__(
        self,
        inner: Any,
        execution: BoundedExecution,
        raise_on_escalation: bool,
    ) -> None:
        self._inner = inner
        self._execution = execution
        self._raise_on_escalation = raise_on_escalation

    def create(self, **kwargs: Any) -> Any:
        self._preflight()
        response = self._inner.create(**kwargs)
        self._account(response)
        return response

    def __getattr__(self, name: str) -> Any:
        return getattr(self._inner, name)

    def _preflight(self) -> None:
        pre = self._execution.check()
        if pre is StepOutcome.LOCKED:
            from iaiso.core import ExecutionLocked
            raise ExecutionLocked(
                f"execution {self._execution.engine.execution_id} is locked"
            )
        if pre is StepOutcome.ESCALATED and self._raise_on_escalation:
            raise EscalationRaised(
                f"execution escalated at pressure "
                f"{self._execution.engine.pressure:.3f}"
            )

    def _account(self, response: Any) -> None:
        tokens = 0
        usage = getattr(response, "usage", None)
        if usage is not None:
            tokens = int(
                getattr(usage, "total_tokens", 0)
                or getattr(usage, "prompt_tokens", 0)
                + getattr(usage, "completion_tokens", 0)
            )

        tool_calls = 0
        choices = getattr(response, "choices", None) or []
        for choice in choices:
            message = getattr(choice, "message", None)
            if message is None:
                continue
            tc = getattr(message, "tool_calls", None) or []
            tool_calls += len(tc)

        model = str(getattr(response, "model", "unknown"))
        self._execution.record_step(
            tokens=tokens,
            tool_calls=tool_calls,
            tag=f"openai.chat.completions.create:{model}",
        )


class AsyncOpenAIBoundedClient:
    """Async variant. Wraps `openai.AsyncOpenAI`."""

    def __init__(
        self,
        client: Any,
        execution: BoundedExecution,
        *,
        raise_on_escalation: bool = False,
    ) -> None:
        self._client = client
        self._execution = execution
        self._raise_on_escalation = raise_on_escalation
        self.chat = _AsyncChatProxy(client.chat, execution, raise_on_escalation)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._client, name)


class _AsyncChatProxy:
    def __init__(
        self,
        inner: Any,
        execution: BoundedExecution,
        raise_on_escalation: bool,
    ) -> None:
        self._inner = inner
        self._execution = execution
        self._raise_on_escalation = raise_on_escalation
        self.completions = _AsyncCompletionsProxy(
            inner.completions, execution, raise_on_escalation
        )

    def __getattr__(self, name: str) -> Any:
        return getattr(self._inner, name)


class _AsyncCompletionsProxy(_CompletionsProxy):
    async def create(self, **kwargs: Any) -> Any:  # type: ignore[override]
        self._preflight()
        response = await self._inner.create(**kwargs)
        self._account(response)
        return response

File: IAIso-v5.0/core/iaiso-node/src/middleware/openai.ts
/**
 * OpenAI middleware.
 *
 * Wraps the `openai` Node SDK so calls to `.chat.completions.create()`
 * and `.responses.create()` (where available) are accounted for in a
 * BoundedExecution. Also works with OpenAI-compatible servers (vLLM,
 * TGI, LiteLLM proxy, Together AI, Groq, etc.) — point the underlying
 * client at the compatible endpoint.
 *
 * Install:
 *   npm install openai
 */

import { BoundedExecution, ExecutionLocked } from "../core/execution.js";
import { StepOutcome } from "../core/types.js";
import { EscalationRaised } from "./anthropic.js";

interface OpenAIChatResponse {
  usage?: {
    prompt_tokens?: number | null;
    completion_tokens?: number | null;
    total_tokens?: number | null;
  } | null;
  choices?: Array<{
    message?: {
      tool_calls?: Array<unknown> | null;
      function_call?: unknown;
    };
  }> | null;
  model?: string;
}

interface OpenAILike {
  chat: {
    completions: {
      create: (params: Record<string, unknown>) => Promise<OpenAIChatResponse> | OpenAIChatResponse;
    };
  };
}

export interface OpenAIBoundedClientOptions {
  raiseOnEscalation?: boolean;
}

export class OpenAIBoundedClient<T extends OpenAILike = OpenAILike> {
  readonly raw: T;
  private readonly _execution: BoundedExecution;
  private readonly _raiseOnEscalation: boolean;
  readonly chat: {
    completions: {
      create: (params: Record<string, unknown>) => Promise<OpenAIChatResponse>;
    };
  };

  constructor(
    client: T,
    execution: BoundedExecution,
    opts: OpenAIBoundedClientOptions = {},
  ) {
    this.raw = client;
    this._execution = execution;
    this._raiseOnEscalation = opts.raiseOnEscalation ?? false;

    const innerCompletions = client.chat.completions;
    const execRef = this._execution;
    const raiseOnEsc = this._raiseOnEscalation;
    const accountFn = this._account.bind(this);

    this.chat = {
      completions: {
        async create(params: Record<string, unknown>): Promise<OpenAIChatResponse> {
          const pre = execRef.check();
          if (pre === StepOutcome.Locked) {
            throw new ExecutionLocked(
              `execution ${execRef.engine.execution_id} is locked`,
            );
          }
          if (pre === StepOutcome.Escalated && raiseOnEsc) {
            throw new EscalationRaised(
              `execution escalated at pressure ${execRef.engine.pressure.toFixed(3)}`,
            );
          }
          const response = await innerCompletions.create(params);
          accountFn(response);
          return response;
        },
      },
    };
  }

  private _account(response: OpenAIChatResponse): void {
    const usage = response.usage ?? null;
    const tokens =
      Number(usage?.total_tokens ?? 0) ||
      Number(usage?.prompt_tokens ?? 0) + Number(usage?.completion_tokens ?? 0);

    let toolCalls = 0;
    const choices = response.choices ?? [];
    for (const choice of choices) {
      const msg = choice?.message;
      if (msg?.tool_calls && Array.isArray(msg.tool_calls)) {
        toolCalls += msg.tool_calls.length;
      } else if (msg?.function_call) {
        toolCalls += 1;
      }
    }

    const model = response.model ?? "unknown";
    this._execution.recordStep({
      tokens,
      tool_calls: toolCalls,
      tag: `openai.chat.completions.create:${model}`,
    });
  }
}
